File: src/main/java/com/example/gawsddns/DDStack.java
package com.example.gawsddns;

import software.amazon.awscdk.Stack;
import software.amazon.awscdk.StackProps;
import software.amazon.awscdk.Duration;
import software.amazon.awscdk.services.lambda.Function;
import software.amazon.awscdk.services.lambda.Runtime;
import software.amazon.awscdk.services.lambda.Code;
import software.amazon.awscdk.services.iam.PolicyStatement;
import software.amazon.awscdk.services.iam.Effect;
import software.amazon.awscdk.services.apigateway.RestApi;
import software.amazon.awscdk.services.apigateway.LambdaIntegration;
import software.amazon.awscdk.services.apigateway.DomainName;
import software.amazon.awscdk.services.apigateway.BasePathMapping;
import software.amazon.awscdk.services.certificatemanager.ICertificate;
import software.amazon.awscdk.services.certificatemanager.Certificate;
import software.amazon.awscdk.services.route53.IHostedZone;
import software.amazon.awscdk.services.route53.HostedZone;
import software.amazon.awscdk.services.route53.HostedZoneProviderProps;
import software.amazon.awscdk.services.route53.ARecord;
import software.amazon.awscdk.services.route53.RecordTarget;
import software.amazon.awscdk.services.route53.targets.ApiGatewayDomain;
import software.constructs.Construct;
import java.util.List;

public class DDStack extends Stack {
    public DDStack(final Construct scope, final String id, final StackProps props) {
        super(scope, id, props);

        Function lambda = Function.Builder.create(this, "DDLambda")
            .runtime(Runtime.JAVA_17)
            .handler("com.example.gawsddns.DDLambda::handleRequest")
            .code(Code.fromAsset("target/gawsddns-1.0-SNAPSHOT.jar"))
            .memorySize(512)
            .timeout(Duration.seconds(10))
            .build();

        lambda.addToRolePolicy(PolicyStatement.Builder.create()
            .actions(List.of("route53:ChangeResourceRecordSets", "route53:ListHostedZones"))
            .resources(List.of("*")) // You can scope this to specific hosted zone
            .effect(Effect.ALLOW)
            .build());

        // Look up your existing wildcard certificate for *.gretrostuff.com
        ICertificate certificate = Certificate.fromCertificateArn(this, "Certificate",
            "arn:aws:acm:us-east-1:223166462382:certificate/b8d54284-3534-4b4f-bac6-726c8fed930a");

        // Look up your existing hosted zone for gretrostuff.com
        IHostedZone hostedZone = HostedZone.fromLookup(this, "HostedZone", 
            HostedZoneProviderProps.builder()
                .domainName("gretrostuff.com")
                .build());

        // Create the custom domain
        DomainName customDomain = DomainName.Builder.create(this, "CustomDomain")
            .domainName("members.gretrostuff.com")
            .certificate(certificate)
            .build();

        // Create the API with direct Lambda integration
        RestApi api = RestApi.Builder.create(this, "DDApi")
            .restApiName("Dynamic DNS API")
            .build();

        // Create Lambda integration with proxy
        LambdaIntegration integration = LambdaIntegration.Builder.create(lambda)
            .proxy(true)  // Enable proxy integration
            .build();

        // Add GET method to root
        api.getRoot().addMethod("GET", integration);
        
        // Add /nic/update path for Dyn compatibility
        api.getRoot()
            .addResource("nic")
            .addResource("update")
            .addMethod("GET", integration);
            
        // Add /v3/update path for Dyn API v3 compatibility
        api.getRoot()
            .addResource("v3")
            .addResource("update")
            .addMethod("GET", integration);

        // Add a proxy resource to catch any other paths
        api.getRoot()
            .addProxy();

        // Map the custom domain to the API for Dyn ddclient compatibility
        BasePathMapping.Builder.create(this, "BasePathMapping")
            .domainName(customDomain)
            .restApi(api)
            .basePath("") // Empty string means root path - Lambda will handle /nic/update routing
            .build();

        // Create DNS record to point members.gretrostuff.com to the custom domain
        ARecord.Builder.create(this, "DDNSRecord")
            .zone(hostedZone)
            .recordName("members")
            .target(RecordTarget.fromAlias(new ApiGatewayDomain(customDomain)))
            .build();
    }
}

File: lib/dd-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as route53 from 'aws-cdk-lib/aws-route53';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as ssm from 'aws-cdk-lib/aws-ssm';
import * as logs from 'aws-cdk-lib/aws-logs';
import { Construct } from 'constructs';

export interface DDStackProps extends cdk.StackProps {
  config: {
    certificateId: string;
    domainName: string;
    subdomainName: string;
    username: string;
    password: string;
  };
}

export class DDStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: DDStackProps) {
    super(scope, id, props);

    const ddLambda = new lambda.Function(this, 'DDLambda', {
      runtime: lambda.Runtime.JAVA_17,
      handler: 'com.example.gawsddns.DDLambda::handleRequest',
      code: lambda.Code.fromAsset('target/gawsddns-1.0-SNAPSHOT.jar'),
      memorySize: 512,
      timeout: cdk.Duration.seconds(30),
    });

    ddLambda.addToRolePolicy(new iam.PolicyStatement({
      actions: ['route53:ChangeResourceRecordSets', 'route53:ListHostedZones', 'route53:ListResourceRecordSets'],
      resources: ['*'],
      effect: iam.Effect.ALLOW,
    }));

    // Look up your existing hosted zone
    const hostedZone = route53.HostedZone.fromLookup(this, 'HostedZone', {
      domainName: props.config.domainName,
    });

    // Create IAM role for API Gateway logging
    const apiGatewayLogRole = new iam.Role(this, 'ApiGatewayLogRole', {
      assumedBy: new iam.ServicePrincipal('apigateway.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AmazonAPIGatewayPushToCloudWatchLogs'),
      ],
    });

    // Create the API Gateway REST API using CfnRestApi
    const api = new apigateway.CfnRestApi(this, 'DDApi', {
      name: 'Dynamic DNS API',
      endpointConfiguration: {
        types: ['REGIONAL'],
        ipAddressType: "dualstack"
      },
    });

    // Set up API Gateway account settings for logging
    new apigateway.CfnAccount(this, 'ApiGatewayAccount', {
      cloudWatchRoleArn: apiGatewayLogRole.roleArn,
    });

    // Create Lambda permission for API Gateway to invoke the Lambda
    new lambda.CfnPermission(this, 'ApiGatewayInvokePermission', {
      action: 'lambda:InvokeFunction',
      functionName: ddLambda.functionName,
      principal: 'apigateway.amazonaws.com',
      sourceArn: `arn:aws:execute-api:${this.region}:${this.account}:${api.ref}/*/*/*`,
    });

    // Create the root resource ("/")
    const rootResource = new apigateway.CfnResource(this, 'RootResource', {
      parentId: api.attrRootResourceId,
      pathPart: '{proxy+}',
      restApiId: api.ref,
    });

    // Create the ANY method with Lambda proxy integration
    const anyMethod = new apigateway.CfnMethod(this, 'AnyMethod', {
      restApiId: api.ref,
      resourceId: rootResource.ref,
      httpMethod: 'ANY',
      authorizationType: 'NONE',
      integration: {
        type: 'AWS_PROXY',
        integrationHttpMethod: 'POST',
        uri: `arn:aws:apigateway:${this.region}:lambda:path/2015-03-31/functions/${ddLambda.functionArn}/invocations`,
      },
      requestParameters: {
        'method.request.path.proxy': true,
      },
    });

    // Deploy the API, depends on method
    const deployment = new apigateway.CfnDeployment(this, 'ApiDeployment', {
      restApiId: api.ref,
      description: 'Deployment without stage for logging config',
    });
    deployment.node.addDependency(anyMethod);

    // Create stage with minimal logging
    const stage = new apigateway.CfnStage(this, 'ApiStage', {
      restApiId: api.ref,
      deploymentId: deployment.ref,
      stageName: 'prod',
      accessLogSetting: {
        destinationArn: `arn:aws:logs:${this.region}:${this.account}:log-group:/aws/apigateway/gawsddns`,
        format: '$context.requestId $context.status $context.error.message $context.integration.error',
      },
    });
    stage.node.addDependency(deployment);

    // Create custom domain with SSL certificate using CfnDomainName
    const certificateArn = `arn:aws:acm:${this.region}:${this.account}:certificate/${props.config.certificateId}`;
    const customDomain = new apigateway.CfnDomainName(this, 'CustomDomain', {
      domainName: `${props.config.subdomainName}.${props.config.domainName}`,
      regionalCertificateArn: certificateArn,
      endpointConfiguration: {
        types: ['REGIONAL'],
        ipAddressType: 'dualstack'
      },
    });

    // Base path mapping for custom domain, depends on domain and deployment
    const basePathMapping = new apigateway.CfnBasePathMapping(this, 'BasePathMapping', {
      domainName: customDomain.domainName!,
      restApiId: api.ref,
      stage: 'prod',
      basePath: '',
    });
    basePathMapping.node.addDependency(customDomain);
    basePathMapping.node.addDependency(stage);

    // Create DNS record for HTTPS endpoint using CfnDomainName outputs
    new route53.ARecord(this, 'DDNSRecord', {
      zone: hostedZone,
      recordName: props.config.subdomainName,
      target: route53.RecordTarget.fromAlias({
        bind: () => ({
          dnsName: customDomain.attrRegionalDomainName,
          hostedZoneId: customDomain.attrRegionalHostedZoneId,
        }),
      }),
    });

    // Create AAAA record for HTTPS endpoint using CfnDomainName outputs
    new route53.AaaaRecord(this, 'DDNSRecordAAAA', {
      zone: hostedZone,
      recordName: props.config.subdomainName,
      target: route53.RecordTarget.fromAlias({
        bind: () => ({
          dnsName: customDomain.attrRegionalDomainName,
          hostedZoneId: customDomain.attrRegionalHostedZoneId,
        }),
      }),
    });

    // Store credentials in SSM Parameter Store
    const usernameParam = new ssm.StringParameter(this, 'DDNSUsernameParam', {
      parameterName: '/gawsddns/username',
      stringValue: props.config.username,
      tier: ssm.ParameterTier.STANDARD
    });
    const passwordParam = new ssm.StringParameter(this, 'DDNSPasswordParam', {
      parameterName: '/gawsddns/password',
      stringValue: props.config.password,
      tier: ssm.ParameterTier.STANDARD
    });

    // Grant Lambda permission to read these parameters
    usernameParam.grantRead(ddLambda);
    passwordParam.grantRead(ddLambda);
  }
}
